### PR TITLE
make cEUR mintable by exchangeEUR

### PR DESCRIPTION
**Pull request by martinvol**
### Description

When doing integration testing with the `ExchangeEUR` and `StableTokenEUR` we discovered that the new Celo Euro exchange was not properly authorized to `mint` and `burn` Celo Euro. 

Because we might want to launch future instantiations of StableToken without waiting on the audit cycle, we want an accommodating design to avoid necessitating source changes. Therefore we decided to parameterize the `StableToken` contract with an `exchangeRegistryId` in storage that can be initialized to `ExchangeEUR`, `ExchangeUSD`, etc. 

### Tested

- Parameterize stabletoken integration tests to test EUR and USD pairings separately
- Added unit tests for `getExchangeRegistryId`

### Related issues

- Fixes #[issue number here]

### Backwards compatibility

- Appends a storage variable to `StableToken` and bumps the `major` version accordingly

### CODE REVIEW

File: packages/protocol/contracts/stability/StableToken.sol [hunk 1]
```diff
@@ -265,7 +271,7 @@ contract StableToken is
    */
   function burn(uint256 value)
     external
-    onlyRegisteredContract(EXCHANGE_REGISTRY_ID)
```
Comment (nambrot) [inline packages/protocol/contracts/stability/StableToken.sol:268]: How will this be backwards compatible for cUSD?

**Comment by asaj** [comment]
PR description please

File: packages/protocol/contracts/stability/StableToken.sol (hunk 1 above)
Comment (yorhodes) [reply packages/protocol/contracts/stability/StableToken.sol:268]: forgot the fallback behavior

Comment (martinvol) [reply packages/protocol/contracts/stability/StableToken.sol:268]: solved here: https://github.com/celo-org/celo-monorepo/pull/6780/files#diff-d3d584105569d60895c3dee70920dfd51d712edd2fef038fa6a5e7e82e3fc18dR397

File: packages/protocol/migrations/09_1_stableToken_EUR.ts [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+    const sortedOracles: SortedOraclesInstance = await getDeployedProxiedContract<
+      SortedOraclesInstance
+    >('SortedOracles', artifacts)
+
+    for (const oracle of config.stableTokenEUR.oracles) {
+      console.info(`Adding ${oracle} as an Oracle for StableToken (EUR)`)
+      await sortedOracles.addOracle(stableToken.address, ensureLeading0x(oracle))
+    }
+
+    const goldPrice = config.stableTokenEUR.goldPrice
```
Comment (codyborn) [inline packages/protocol/migrations/09_1_stableToken_EUR.ts:59]: should we rename this to `celoPrice`?

Comment (martinvol) [reply packages/protocol/migrations/09_1_stableToken_EUR.ts:59]: we can leave it consistent with the rest of the codebase

**Review by kevjue — COMMENTED** [review]
I only took a look at `StableToken.sol` and `release3.json`, and they both look good to me.  However, I'm going to defer to others to approve who have more experience with the core smart contracts.

File: packages/protocol/migrations/10_1_exchange_EUR.ts [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+  ReserveInstance,
+  StableTokenEURInstance,
+} from 'types'
+
+const initializeArgs = async (): Promise<any[]> => {
+  const stableTokenEUR: StableTokenEURInstance = await getDeployedProxiedContract<
+    StableTokenEURInstance
+  >('StableTokenEUR', artifacts)
+  return [
+    // TODO replace this with specific
```
Comment (codyborn) [inline packages/protocol/migrations/10_1_exchange_EUR.ts]: What's this TODO for? Should we create a ticket? (channeling @nambrot 😄)

Comment (yorhodes) [reply packages/protocol/migrations/10_1_exchange_EUR.ts]: cc @martinvol but I think we can simply remove

File: packages/protocol/test/common/integration.ts [hunk 4]
```diff
@@ -382,180 +382,185 @@ contract('Integration: Governance', (accounts: string[]) => {
...
+          originalReserve = await goldToken.balanceOf(reserve.address)
+          await goldToken.approve(exchange.address, maxSellAmount)
+          await exchange.buy(buyAmount, maxSellAmount, false)
+          finalStable = await stableToken.balanceOf(accounts[0])
+          finalGold = await goldToken.balanceOf(accounts[0])
+          finalReserve = await goldToken.balanceOf(reserve.address)
+        })
+
+        it(`should increase user's stable`, async () => {
+          assert.isTrue(finalStable.gt(originalStable))
```
Comment (codyborn) [inline packages/protocol/test/common/integration.ts:492]: Can we test that it's greater than by the expected amount? (similar for the rest of the tests)

**Review by codyborn — COMMENTED** [review]
€!

File: packages/protocol/test/common/integration.ts (hunk 4 above)
Comment (yorhodes) [reply packages/protocol/test/common/integration.ts:492]: These tests were not written for this PR. If we want to improve the correctness we can save this for a followup PR. In the spirit of getting this release to the auditors asap, I merged to `v3.pre-audit`.

File: packages/protocol/contracts/stability/StableToken.sol [hunk 5]
```diff
@@ -67,6 +67,9 @@ contract StableToken is
 
   InflationState inflationState;
 
+  // for differentiating between corresponding fiat pairs
```
Comment (asaj) [inline packages/protocol/contracts/stability/StableToken.sol:70]: nit: `The registry ID of the exchange contract with permission to mint and burn this token. Unique per StableToken instance.`

File: packages/protocol/contracts/stability/StableToken.sol [hunk 6]
```diff
@@ -383,6 +389,19 @@ contract StableToken is
     return _valueToUnits(updatedInflationFactor, value);
   }
 
+  /**
+   * @notice Returns the exchange id in the registry of the corresponding fiat pair exchange.
+   * @dev When this storage is uninitialized, it falls back to the default EXCHANGE_REGISTRY_ID.
```
Comment (asaj) [inline packages/protocol/contracts/stability/StableToken.sol:394]: Can we include a comment about *why* this is the case? And under what conditions we might deprecate this default value?

File: packages/protocol/migrations/09_1_stableToken_EUR.ts [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+    rate.toString(),
+    config.stableTokenEUR.inflationPeriod,
+    config.stableTokenEUR.initialBalances.addresses,
+    config.stableTokenEUR.initialBalances.values,
+    'ExchangeEUR',
+  ]
+}
+
+// TODO make this general
+module.exports = deploymentForCoreContract<StableTokenEURInstance>( // TODO add EUR
```
Comment (asaj) [inline packages/protocol/migrations/09_1_stableToken_EUR.ts:37]: Is this TODO still needed?

File: packages/protocol/migrations/10_1_exchange_EUR.ts [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+        artifacts
+      )
+      await freezer.freeze(exchange.address)
+    }
+
+    const reserve: ReserveInstance = await getDeployedProxiedContract<ReserveInstance>(
+      'Reserve',
+      artifacts
+    )
+    // cUSD doesn't need to be added as it currently harcoded on reserve.sol
```
Comment (asaj) [inline packages/protocol/migrations/10_1_exchange_EUR.ts:49]: But don't we want to add it anyway so we can remove this bit of messiness from Reserve.sol?

Comment (asaj) [inline packages/protocol/migrations/10_1_exchange_EUR.ts:49]: nit: `Reserve.sol`.

File: packages/protocol/releaseData/initializationData/release3.json [hunk 9]
```diff
@@ -1,4 +1,4 @@
 {
-  "StableTokenEUR": ["Celo Euro", 18, "0x000000000000000000000000000000000000ce10", "1", "47304000", "0x0000000000000000000000000000000000000000", "0"],
+  "StableTokenEUR": ["Celo Euro", "cEUR", 18, "0x000000000000000000000000000000000000ce10", "1", "47304000", "0x0000000000000000000000000000000000000000", "0", "ExchangeEUR"],
```
Comment (asaj) [inline packages/protocol/releaseData/initializationData/release3.json:2]: I'm assuming most of the parameters in this file are placeholders? Too bad json doesn't allow comments :/

File: packages/protocol/migrations/10_1_exchange_EUR.ts (hunk 8 above)
Comment (tkporter) [reply packages/protocol/migrations/10_1_exchange_EUR.ts:49]: Sounds like CR3 will not involve cUSD's exchange being added as an exchange spender (correct me if I'm wrong), in that case do you think it makes sense for our migrations to match mainnet and not have cUSD's exchange as spender?

**Pull request merged**